File: frontend/types/resume.ts
export interface ContactInfo {
  full_name: string;
  email: string;
  phone: string;
  location: string;
}

export interface LinkItem {
  id: string;
  label: string;
  url: string;
}

export interface ExperienceEntry {
  id: string;
  company: string;
  title: string;
  location: string;
  start_date: string;
  end_date: string;
  bullets: string[];
}

export interface EducationEntry {
  id: string;
  institution: string;
  degree: string;
  location: string;
  start_date: string;
  end_date: string;
  gpa: string;
}

export interface ProjectEntry {
  id: string;
  name: string;
  url: string;
  bullets: string[];
}

export interface SkillCategory {
  id: string;
  name: string;
  skills: string[];
}

export interface ResumeContent {
  contact: ContactInfo;
  links: LinkItem[];
  summary: string;
  experience: ExperienceEntry[];
  education: EducationEntry[];
  projects: ProjectEntry[];
  skills: SkillCategory[];
}

export interface ResumeDocument {
  id: string;
  title: string;
  job_title: string | null;
  job_url: string | null;
  status: string;
  job_description: string | null;
  company_url: string | null;
  company_name: string | null;
  source_type: string;
  content_json: ResumeContent;
  latex_source: string | null;
  insights_json: Record<string, unknown> | null;
  create_cover_letter: boolean;
  cover_letter_meta: Record<string, string> | null;
  application_id: string | null;
  job_id: string | null;
  inbox_job_id: string | null;
  resume_category: string | null;
  why_this_version: {
    category?: string;
    category_source?: string;
    category_confidence?: number | null;
    matched_keywords?: string[];
    missing_keywords?: string[];
    fit_score?: number | null;
    truthfulness?: string;
    template_notes?: Record<string, unknown>;
  } | null;
  cover_letter_id: string | null;
  pipeline_error?: string | null;
  last_step?: string | null;
  created_at: string;
  updated_at: string;
}

export interface CoverLetterContent {
  recipient_name: string;
  recipient_title: string;
  company_name: string;
  company_address: string;
  date: string;
  salutation: string;
  paragraphs: string[];
  closing: string;
}

export function emptyCoverLetterContent(): CoverLetterContent {
  return {
    recipient_name: "",
    recipient_title: "",
    company_name: "",
    company_address: "",
    date: "",
    salutation: "Dear Hiring Manager,",
    paragraphs: [""],
    closing: "Sincerely,",
  };
}

export function parseCoverLetterContent(raw: Record<string, unknown>): CoverLetterContent {
  const base = emptyCoverLetterContent();
  const paragraphs = raw.paragraphs;
  return {
    ...base,
    recipient_name: String(raw.recipient_name || ""),
    recipient_title: String(raw.recipient_title || ""),
    company_name: String(raw.company_name || ""),
    company_address: String(raw.company_address || ""),
    date: String(raw.date || ""),
    salutation: String(raw.salutation || base.salutation),
    closing: String(raw.closing || base.closing),
    paragraphs: Array.isArray(paragraphs) && paragraphs.length
      ? paragraphs.map((p) => String(p))
      : [""],
  };
}

export interface CoverLetterDocument {
  id: string;
  title: string;
  status: string;
  resume_id: string | null;
  hiring_manager_name: string | null;
  hiring_manager_email: string | null;
  street_address: string | null;
  city: string | null;
  state_province: string | null;
  postal_code: string | null;
  letter_date: string | null;
  additional_context: string | null;
  content_json: Record<string, unknown>;
  latex_source: string | null;
  created_at: string;
  updated_at: string;
}

export interface CoverLetterMeta {
  hiring_manager_name?: string;
  hiring_manager_email?: string;
  street_address?: string;
  city?: string;
  state_province?: string;
  postal_code?: string;
  letter_date?: string;
  additional_context?: string;
}

export interface ApiKeyConfig {
  id: string;
  provider: string;
  api_key_masked: string;
  base_url: string | null;
  model_name: string | null;
  embedding_model: string | null;
  is_default: boolean;
}

export interface ApiToken {
  id: string;
  name: string;
  token_prefix: string;
  created_at: string;
  last_used_at: string | null;
}

export interface PendingChange {
  id: string;
  path: string;
  path_label?: string | null;
  old_value: string | null;
  new_value: string | null;
  status: string;
}

export interface ChatMessage {
  id: string;
  role: string;
  content: string;
  pending_changes: PendingChange[];
  created_at: string;
}

export interface ChatExchange {
  user_message: ChatMessage;
  assistant_message: ChatMessage;
}

export interface ATSSuggestionItem {
  text: string;
  prompt: string;
  priority: string;
  category: string;
}

export interface ATSScore {
  id: string;
  overall_score: number;
  keyword_match: number;
  formatting_score: number;
  semantic_score?: number;
  skills_coverage?: number;
  section_score?: number;
  matched_keywords?: string[] | null;
  missing_keywords: string[] | null;
  suggestions: string[];
  suggestion_items?: ATSSuggestionItem[];
  breakdown?: Record<string, unknown> | null;
  created_at: string;
}

export interface ResumeStatus {
  id: string;
  status: string;
  last_step?: string | null;
  pipeline_error?: string | null;
  cover_letter_id?: string | null;
  ats_score?: ATSScore | null;
  updated_at: string;
}

export interface StructuredProfile {
  content: ResumeContent;
  updated_at?: string;
}

export function emptyResumeContent(): ResumeContent {
  return {
    contact: { full_name: "", email: "", phone: "", location: "" },
    links: [],
    summary: "",
    experience: [],
    education: [],
    projects: [],
    skills: [],
  };
}

export function newId(): string {
  return crypto.randomUUID();
}

File: backend/app/models/resume.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, String, Text, func
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class ResumeDocument(Base):
    __tablename__ = "resume_documents"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"))
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    status: Mapped[str] = mapped_column(String(20), default="draft")
    job_description: Mapped[str | None] = mapped_column(Text)
    company_url: Mapped[str | None] = mapped_column(String(500))
    company_name: Mapped[str | None] = mapped_column(String(255))
    source_type: Mapped[str] = mapped_column(String(20), default="profile")
    content_json: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    latex_source: Mapped[str | None] = mapped_column(Text)
    insights_json: Mapped[dict | None] = mapped_column(JSONB)
    application_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("user_applications.id", ondelete="SET NULL")
    )
    job_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("jobs.id", ondelete="SET NULL"), index=True
    )
    inbox_job_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inbox_jobs.id", ondelete="SET NULL"), index=True
    )
    resume_category: Mapped[str | None] = mapped_column(String(50))
    why_this_version: Mapped[dict | None] = mapped_column(JSONB)
    create_cover_letter: Mapped[bool] = mapped_column(Boolean, default=False)
    cover_letter_meta: Mapped[dict | None] = mapped_column(JSONB)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    user = relationship("User", back_populates="resumes")
    cover_letter = relationship("CoverLetterDocument", back_populates="resume", foreign_keys="CoverLetterDocument.resume_id")
    agent_runs = relationship("AgentRun", back_populates="resume", cascade="all, delete-orphan")
    chat_messages = relationship("ChatMessage", back_populates="resume", cascade="all, delete-orphan")
    ats_scores = relationship("ATSScore", back_populates="resume", cascade="all, delete-orphan")


class AgentRun(Base):
    __tablename__ = "agent_runs"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    resume_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("resume_documents.id", ondelete="CASCADE")
    )
    agent_type: Mapped[str] = mapped_column(String(50), nullable=False)
    status: Mapped[str] = mapped_column(String(20), default="pending")
    steps_json: Mapped[dict | None] = mapped_column(JSONB)
    error: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    resume = relationship("ResumeDocument", back_populates="agent_runs")


class ChatMessage(Base):
    __tablename__ = "chat_messages"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    resume_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("resume_documents.id", ondelete="CASCADE")
    )
    role: Mapped[str] = mapped_column(String(20), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    pending_diffs_json: Mapped[list | None] = mapped_column(JSONB)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    resume = relationship("ResumeDocument", back_populates="chat_messages")
    pending_changes = relationship("PendingChange", back_populates="message", cascade="all, delete-orphan")


class PendingChange(Base):
    __tablename__ = "pending_changes"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    message_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("chat_messages.id", ondelete="CASCADE")
    )
    path: Mapped[str] = mapped_column(String(500), nullable=False)
    old_value: Mapped[str | None] = mapped_column(Text)
    new_value: Mapped[str | None] = mapped_column(Text)
    status: Mapped[str] = mapped_column(String(20), default="pending")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    message = relationship("ChatMessage", back_populates="pending_changes")


class ATSScore(Base):
    __tablename__ = "ats_scores"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    resume_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("resume_documents.id", ondelete="CASCADE")
    )
    job_description: Mapped[str | None] = mapped_column(Text)
    overall_score: Mapped[int] = mapped_column(default=0)
    keyword_match: Mapped[int] = mapped_column(default=0)
    formatting_score: Mapped[int] = mapped_column(default=0)
    semantic_score: Mapped[int] = mapped_column(default=0)
    skills_coverage: Mapped[int] = mapped_column(default=0)
    section_score: Mapped[int] = mapped_column(default=0)
    matched_keywords: Mapped[list | None] = mapped_column(JSONB)
    suggestions_json: Mapped[dict | None] = mapped_column(JSONB)
    breakdown_json: Mapped[dict | None] = mapped_column(JSONB)
    missing_keywords: Mapped[list | None] = mapped_column(JSONB)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    resume = relationship("ResumeDocument", back_populates="ats_scores")
